Zynos Discovery and Polling Enhancements

Jimmy-Cl opened the pull request:
Please give a short description what your pull request is for

Enhancements for Zynos OS -
- Human Readable port numbers
- Added Dbm levels to ports
- Added Stacking state sensor

<img width="919" alt="image" src="https://user-images.githubusercontent.com/68513176/152556538-3a76d75e-9415-461c-b9a5-ed9604782b4e.png">

<img width="1894" alt="image" src="https://user-images.githubusercontent.com/68513176/152556695-a0b473b6-97db-4207-bf93-8de3bbcb76dc.png">

<img width="930" alt="image" src="https://user-images.githubusercontent.com/68513176/152556611-5773b8ba-33aa-409e-bf4a-5c78439c9a80.png">


DO NOT DELETE THE UNDERLYING TEXT

#### Please note

> Please read this information carefully. You can run `./lnms dev:check` to check your code before submitting.

- [x] Have you followed our [code guidelines?](https://docs.librenms.org/Developing/Code-Guidelines/)
- [x] If my Pull Request does some changes/fixes/enhancements in the WebUI, I have inserted a screenshot of it.
- [x] If my Pull Request makes discovery/polling/yaml changes, I have added/updated [test data](https://docs.librenms.org/Developing/os/Test-Units/).

#### Testers

If you would like to test this pull request then please run: `./scripts/github-apply <pr_id>`, i.e `./scripts/github-apply 5926`
After you are done testing, you can remove the changes with `./scripts/github-remove`.  If there are schema changes, you can ask on discord how to revert.

PipoCanaja commented:
Hi @Jimmy-Cl 
Could you check the existing test JSON data and re-generate them to take into account your improvements ? This is needed so the regression tests succeed.

`tests/data/zynos.json`:
https://github.com/librenms/librenms/runs/5145911327?check_suite_focus=true

Jimmy-Cl commented:
@PipoCanaja I have updated the zynos.json and zynos.snmprec files so the tests should pass now.

PipoCanaja commented on includes/definitions/discovery/zynos.yaml:
```diff
@@ -26,6 +26,61 @@ modules:
+            data:
+                -
+                    oid:
+                        - ifDescr
```
ifDescr is not necessary here, unused.

Jimmy-Cl replied:
Well spotted, I originally used ifDescr but must have forgotten to remove it from the pre-cache

PipoCanaja commented on includes/definitions/discovery/zynos.yaml:
```diff
@@ -26,6 +26,61 @@ modules:
+                -
+                    oid:
+                        - ifDescr
+                        - zyStackingSlotInfoStackingStatus
```
Almost sure this one is not necessary, as it is part of zyxelStackingSlotInfoTable so it will be already loaded. 
```suggestion
                        - zyStackingSlotInfoStackingStatus
```

Jimmy-Cl replied:
Once again forgot to remove the pre-cache after I stopped using zyStackingSlotInfoStackingStatus with different tables.

PipoCanaja commented on tests/snmpsim/zynos.snmprec:
```diff
@@ -64,66 +58,30 @@
-1.0.8802.1.1.2.1.3.7.1.3.428|4|220
-1.0.8802.1.1.2.1.3.7.1.3.429|4|221
-1.0.8802.1.1.2.1.3.7.1.3.430|4|222
+1.0.8802.1.1.2.1.4.1.1.4.120.128.1408|2|4
```
You are not supposed to change an existing snmprec file. Only create new ones. And (re) calculate the JSON files. 
SNMPREC is a capture of the device, and JSON is the calculated as a "reference" on how LibreNMS sees the device.
And regression test will be compared again at every change of the source code  to detect any issue.

Jimmy-Cl replied:
I have reverted my zynos.snmprec changes. Can you just confirm that what I did to the JSON files was correct, as I ran a new test and renamed it to be zynos.json not zynos_xgs4600.json? It would not let me run a test without a variant. If this is wrong and you have another script you want me to run just let me know.

laf commented:
It's showing as the `zynos.json` being modified. If you take the contents of this file and save it as `zynos.json` that will restore it, then commit the file and push here.

Jimmy-Cl commented:
I have restored zynos.json as you wanted however the tests will most likley still fail as the changes I have made effect the ifDescr, ifName and ifAlias of zynos switch ports. When your tests run they use the existing zynos.json not the zynos_xgs4600.json file that I generated so they show an error which is, OS zynos: Discovered ports data does not match that found in tests/data/zynos.json. Can we simply ignore this error or is there a way of me regenerating the existing zynos.json file to have my new port fDescr, ifName and ifAlias.

murrant commented on includes/discovery/ports/zynos.inc.php:
```diff
@@ -0,0 +1,60 @@
+        if ($portNum < 50) {
+            $portNum++;
+            //Leading 0 for single digits
+            if ($portNum < 10) {
```
instead of doing that, you can just do this:
```php
$ifDescr = '1/' . zeropad($portNum);
```

Jimmy-Cl replied:
Never used zeropad but it looks interesting, I will test it and get back to you.

Jimmy-Cl replied:
Zeropad works really well, Thanks for pointing this out :)

murrant commented on includes/discovery/ports/zynos.inc.php:
```diff
@@ -0,0 +1,60 @@
+    }
+    //set the libre ifDescr value to new port name
+    $port_stats[$index]['ifDescr'] = $ifDescr;
+    $port_stats[$index]['ifName'] = $ifDescr;
```
It might be worthwhile to only override ifName.  if ifDescr is empty, you could set it to the original ifName value.
ifDescr is the port as described by the os
ifName should be the name of the port as used in cli commands
ifAlias is the human configured description of the port

murrant replied:
There is also an os definition flag to set which field to prefer for displaying the port name.

Jimmy-Cl replied:
We tested the OS definition flag but the reason that we changed both ifDescr and ifName was because some areas of libre would still show as swp numbers, The cli uses the 1/1 format and so does the writing on the switch so thats another reson why we wanted to change ifDescr. I will test it again to see which areas use ifDescr and if it is still necessary.

Jimmy-Cl replied:
After testing it would be best if we can change both ifDescr and ifName so we can completely remove the swp numbers. If you guys don't like us changing ifDescr let me know and I will remove it. Please can someone rerun the workflows now I have added the zeropad config so it can test the pull request.

murrant commented:
We can't merge it without succeeding tests.
If you changed the expected result(data) then you will need to update the LibreNMS test data to reflect that new data.
[./scripts/save-test-data.php ](https://docs.librenms.org/Developing/os/Test-Units/#2-save-test-data)

Jimmy-Cl commented:
@murrant I have updated my pull request so that it doesn't change ifDescr like you requested. I have added ifname: true to the zynos.yaml file so that ifName overrides the ifDescr for the name of the port. I have also included new test data which confirms that ifDescr stays as the swp01 format and confirmed on our dev instance that the ports on the front end change to be the new ifName like we want. Let me know if you spot any issues.

Jimmy-Cl commented:
I have moved my Re-Generated Zynos.json to this pull request please can someone re-run the tests below.

Jimmy-Cl commented:
Closing this pull request and changing my script so it can pass the tests
